src/module/producers.rs

```diff
@@ -0,0 +1,93 @@
+//! Handling of the wasm `producers` section
+//!
+//! Specified upstream at
+//! https://github.com/WebAssembly/tool-conventions/blob/master/ProducersSection.md
+
+use crate::error::Result;
+use failure::bail;
+use parity_wasm::elements::*;
+use std::collections::HashMap;
+
+/// Representation of the wasm custom section `producers`
+#[derive(Debug, Default)]
+pub struct ModuleProducers {
+    fields: HashMap<String, HashMap<String, String>>,
+}
+
+impl ModuleProducers {
+    /// Parse a producers section from the custom section payload specified.
+    pub fn parse(mut data: &[u8]) -> Result<ModuleProducers> {
+        let mut ret = ModuleProducers::default();
+        let amt: u32 = VarUint32::deserialize(&mut data)?.into();
+
+        for _ in 0..amt {
+            let name = String::deserialize(&mut data)?;
+            let cnt: u32 = VarUint32::deserialize(&mut data)?.into();
+
+            let map = ret.fields.entry(name).or_insert(HashMap::new());
+            for _ in 0..cnt {
+                let name = String::deserialize(&mut data)?;
+                let version = String::deserialize(&mut data)?;
+                map.insert(name, version);
+            }
+        }
+        if data.len() != 0 {
+            bail!("failed to decode all data in producers section");
+        }
+
+        Ok(ret)
+    }
+
+    /// Serialize this producers section into its binary format
+    pub fn emit(&self) -> Vec<u8> {
+        // re-serialize these fields back into the custom section
+        let mut dst = Vec::new();
+        VarUint32::from(self.fields.len() as u32)
+            .serialize(&mut dst)
+            .unwrap();
+
+        // emit in a deterministic order, so be sure to sort
+        let mut fields = self.fields.iter().collect::<Vec<_>>();
+        fields.sort_by_key(|f| f.0);
```

Comment:
Hm... if we maintained the input order, then that would also be deterministic. And if every tool only appends and maintains input order, then you get a picture of which tools processed the wasm binary in what order. I think it is worth trying to maintain that.